Documentation: Add notes about subtypes of concrete types (#58017)
JuliaLang/julia: doc/src/manual/types.md

While this might at first seem unduly restrictive, it has many beneficial consequences with surprisingly
few drawbacks. It turns out that being able to inherit behavior is much more important than being
able to inherit structure, and inheriting both causes significant difficulties in traditional

> object-oriented languages. Despite these restrictions, concrete types may have abstract subtypes, but
> there are only few examples of this (see e.g. [Abstract Types](@ref man-abstract-types)) and subtypes
> of concrete types cannot be declared manually inside the language.
> 
> Other high-level aspects of Julia's type system that should be mentioned up front are:

Reviewer: Despite the statement "subtypes of concrete types cannot be declared manually inside the language.", The language "Despite these restrictions, concrete types may have abstract subtypes" makes me think that there is some user choice about this subtyping (and that some concrete types have no subtypes).

Perhaps this language would be better to clarify that these subtypings are not within user control and are always present in the case of `Union{}`:

```suggestion
object-oriented languages. While concrete types do have abstract subtypes, there are only two examples of this
([`Union{}`](@ref man-abstract-types) and [`Type{T}`](@ref man-typet-type))) and additional subtypes
of concrete types cannot be declared.

Other high-level aspects of Julia's type system that should be mentioned up front are:
```

>

Author: Thanks for the suggestion! I find it clearer the way you phrased it.

About the other sentence: I thought so, but didn't know for sure if they are the only two examples.

called "top" because it is at the apex of the type graph. Julia also has a predefined abstract
"bottom" type, at the nadir of the type graph, which is written as `Union{}`. It is the exact
opposite of `Any`: no object is an instance of `Union{}` and all types are supertypes of `Union{}`.

> This also means that concrete types formally have `Union{}` as a (abstract) subtype.

Author: Agreed

WrapType{Type{Float64}}(Float64)
```

> This behavior of `Type{Float64}` is another example of an abstract type subtyping a

Author: Agreed. I'm also not sure if this "cross-reference" is really necessary here, but since the page is quite long I thought a reminder at this place would be helpful. But "another" really made it sound like there is some reference intended.
